In [72]:
from IPython.display import display

### Overview of the Data Mining Process

- Define the purpose
- Obtain the data
- Explore and clean the data
- Determine the Data mining task
- Choose the Data Mining methods
- Apply the methods and select the final model
- Evaluate the performance
- Deploy

For our first project we will use the West Roxbury, Boston housing data to predict the home values.

First lets load the data.

In [1]:
import pandas as pd

In [2]:
housing_df = pd.read_csv('dmba-datasets\dmba\WestRoxbury.csv')

We will first explore the column names used in the data set.

In [3]:
housing_df.columns.tolist()

['TOTAL VALUE ',
 'TAX',
 'LOT SQFT ',
 'YR BUILT',
 'GROSS AREA ',
 'LIVING AREA',
 'FLOORS ',
 'ROOMS',
 'BEDROOMS ',
 'FULL BATH',
 'HALF BATH',
 'KITCHEN',
 'FIREPLACE',
 'REMODEL']

Lets us look at the dimensions of our data set.

In [4]:
housing_df.shape

(5802, 14)

The data has 5802 rows and 14 columns as we saw above.

Let us look at the first few rows of the data.

In [5]:
housing_df.head()

TOTAL VALUE    TAX  LOT SQFT   YR BUILT  GROSS AREA   LIVING AREA  FLOORS   \
0         344.2  4330       9965      1880         2436         1352      2.0   
1         412.6  5190       6590      1945         3108         1976      2.0   
2         330.1  4152       7500      1890         2294         1371      2.0   
3         498.6  6272      13773      1957         5032         2608      1.0   
4         331.5  4170       5000      1910         2370         1438      2.0   

   ROOMS  BEDROOMS   FULL BATH  HALF BATH  KITCHEN  FIREPLACE REMODEL  
0      6          3          1          1        1          0    None  
1     10          4          2          1        1          0  Recent  
2      8          4          1          1        1          0    None  
3      9          5          1          1        1          1    None  
4      7          3          2          0        1          0    None

We can see that the first house has a total value of \\$344.2K and paid a tax amount of \$4330. It had a lot size of 9965 and was built in the year 1880.

The column names include a space, we will remove the space and replace it with an `_`(underscore).

In [6]:
housing_df.columns = [x.strip().replace(' ', '_') for x in housing_df.columns]

We can see below that we have the column names in the desired format.

In [7]:
housing_df.head()

TOTAL_VALUE   TAX  LOT_SQFT  YR_BUILT  GROSS_AREA  LIVING_AREA  FLOORS  \
0        344.2  4330      9965      1880        2436         1352     2.0   
1        412.6  5190      6590      1945        3108         1976     2.0   
2        330.1  4152      7500      1890        2294         1371     2.0   
3        498.6  6272     13773      1957        5032         2608     1.0   
4        331.5  4170      5000      1910        2370         1438     2.0   

   ROOMS  BEDROOMS  FULL_BATH  HALF_BATH  KITCHEN  FIREPLACE REMODEL  
0      6         3          1          1        1          0    None  
1     10         4          2          1        1          0  Recent  
2      8         4          1          1        1          0    None  
3      9         5          1          1        1          1    None  
4      7         3          2          0        1          0    None

Finally we can look at the important statistics on each of our columns.

In [8]:
housing_df.describe()

TOTAL_VALUE           TAX      LOT_SQFT     YR_BUILT   GROSS_AREA  \
count  5802.000000   5802.000000   5802.000000  5802.000000  5802.000000   
mean    392.685715   4939.485867   6278.083764  1936.744916  2924.842123   
std      99.177414   1247.649118   2669.707974    35.989910   883.984726   
min     105.000000   1320.000000    997.000000     0.000000   821.000000   
25%     325.125000   4089.500000   4772.000000  1920.000000  2347.000000   
50%     375.900000   4728.000000   5683.000000  1935.000000  2700.000000   
75%     438.775000   5519.500000   7022.250000  1955.000000  3239.000000   
max    1217.800000  15319.000000  46411.000000  2011.000000  8154.000000   

       LIVING_AREA       FLOORS        ROOMS     BEDROOMS    FULL_BATH  \
count  5802.000000  5802.000000  5802.000000  5802.000000  5802.000000   
mean   1657.065322     1.683730     6.994829     3.230093     1.296794   
std     540.456726     0.444884     1.437657     0.846607     0.522040   
min     504.000000     1.000000     3.000000     1.000000     1.000000   
25%    1308.000000     1.000000     6.000000     3.000000     1.000000   
50%    1548.500000     2.000000     7.000000     3.000000     1.000000   
75%    1873.750000     2.000000     8.000000     4.000000     2.000000   
max    5289.000000     3.000000    14.000000     9.000000     5.000000   

         HALF_BATH     KITCHEN    FIREPLACE  
count  5802.000000  5802.00000  5802.000000  
mean      0.613926     1.01534     0.739917  
std       0.533839     0.12291     0.565108  
min       0.000000     1.00000     0.000000  
25%       0.000000     1.00000     0.000000  
50%       1.000000     1.00000     1.000000  
75%       1.000000     1.00000     1.000000  
max       3.000000     2.00000     4.000000

#### Sampling from a database

For any data mining project, we perform our analysis on less than complete data. Typically we sample a small subset of data. We can sample data using the below command.

In [9]:
# Getting a sample size of 5
housing_df.sample(5)

TOTAL_VALUE   TAX  LOT_SQFT  YR_BUILT  GROSS_AREA  LIVING_AREA  FLOORS  \
5618        624.3  7853      6886      1953        5272         3192     2.0   
5276        345.8  4350      6126      1953        2494         1341     1.5   
818         333.1  4190     11400      1920        3188         1958     2.0   
1511        328.1  4127      4147      1929        3410         1295     1.0   
4862        314.6  3957      3410      1969        1907         1080     2.0   

      ROOMS  BEDROOMS  FULL_BATH  HALF_BATH  KITCHEN  FIREPLACE REMODEL  
5618      9         5          2          0        1          1  Recent  
5276      7         3          1          1        1          1    None  
818       8         3          2          0        1          0    None  
1511      5         2          1          0        1          1  Recent  
4862      6         3          1          1        1          1    None

Sometimes we may want to oversample from a certain category of data. We can assign higher weights to such categories and use the sample method again.

Below we will over sample the data to select houses built before the year 1920.

In [10]:
weights = [0.9 if year < 1920 else 0.01 for year in housing_df['YR_BUILT']]
housing_df.sample(10, weights = weights)

TOTAL_VALUE   TAX  LOT_SQFT  YR_BUILT  GROSS_AREA  LIVING_AREA  FLOORS  \
2709        428.2  5386      7500      1910        3132         1923     2.0   
2855        484.3  6092     10569      1890        4653         2412     2.0   
2985        482.8  6073      5096      1906        4724         2674     2.0   
4666        419.2  5273      5590      1910        3597         2151     2.0   
3442        559.0  7032      5532      1892        4298         2525     2.5   
3685        458.6  5769      9600      1896        3356         2078     2.0   
2390        391.2  4921      4002      1910        2811         1732     2.0   
3001        387.8  4878      7500      1887        2796         1648     2.0   
3718        730.9  9194     22956      1900        4500         2909     2.5   
3263        364.2  4581     11168      1915        2944         1664     2.0   

      ROOMS  BEDROOMS  FULL_BATH  HALF_BATH  KITCHEN  FIREPLACE REMODEL  
2709      7         4          1          1        1          1    None  
2855      8         3          1          1        1          1    None  
2985      8         4          1          1        1          1    None  
4666      8         3          1          1        1          1    None  
3442      9         5          4          1        1          1  Recent  
3685      7         3          2          1        1          0     Old  
2390      7         4          1          1        1          0    None  
3001      7         3          1          1        1          1    None  
3718     11         6          3          1        1          0  Recent  
3263      7         3          1          1        1          0    None

#### Processing and Cleaning the data.

There are different types of variables as shown below.

- Numerical
    - Discrete
        - Eg Number of students in a class
    - Continuous
        - Eg Daily Temperature
- Characters (Strings)
    - Categorical
        - Nominal (No order)
            - Male versus Female
        - Ordinal (Ordered data)
            - Movie Ratings (1 star vs 5 stars)

In our housing data, we can see that the remodel data has string characters. We can observe that this data type is a categorical data. We need to convert it to that.

In [13]:
housing_df['REMODEL'] = housing_df['REMODEL'].astype('category')

In [14]:
housing_df['REMODEL'].dtypes

CategoricalDtype(categories=['None', 'Old', 'Recent'], ordered=False)

#### Creating Dummy variables in pandas

We will convert our categorical data into numerical dummies.

In [18]:
housing_df = pd.get_dummies(housing_df,prefix_sep='_', drop_first=True)
housing_df.columns

Index(['TOTAL_VALUE', 'TAX', 'LOT_SQFT', 'YR_BUILT', 'GROSS_AREA',
       'LIVING_AREA', 'FLOORS', 'ROOMS', 'BEDROOMS', 'FULL_BATH', 'HALF_BATH',
       'KITCHEN', 'FIREPLACE', 'REMODEL_Old', 'REMODEL_Recent'],
      dtype='object')

In [20]:
housing_df.loc[:, ['REMODEL_Old', 'REMODEL_Recent']].head()

REMODEL_Old  REMODEL_Recent
0            0               0
1            0               1
2            0               0
3            0               0
4            0               0

#### Normalizing data

Some machine learning algorithms require the data to be normalized. Below we will look a the three methods to normalize the data.

In [26]:
# Method 1

norm_df = (housing_df - housing_df.mean()) / housing_df.std()
norm_df.head()

TOTAL_VALUE       TAX  LOT_SQFT  YR_BUILT  GROSS_AREA  LIVING_AREA  \
0    -0.488879 -0.488507  1.381019 -1.576690   -0.552998    -0.564458   
1     0.200795  0.200789  0.116835  0.229372    0.207196     0.590121   
2    -0.631048 -0.631176  0.457697 -1.298834   -0.713635    -0.529303   
3     1.067927  1.068020  2.807392  0.562799    2.383704     1.759502   
4    -0.616932 -0.616749 -0.478735 -0.743123   -0.627660    -0.405334   

     FLOORS     ROOMS  BEDROOMS  FULL_BATH  HALF_BATH   KITCHEN  FIREPLACE  \
0  0.710905 -0.691980 -0.271783  -0.568528   0.723202 -0.124803  -1.309337   
1  0.710905  2.090325  0.909403   1.347035   0.723202 -0.124803  -1.309337   
2  0.710905  0.699173  0.909403  -0.568528   0.723202 -0.124803  -1.309337   
3 -1.536872  1.394749  2.090589  -0.568528   0.723202 -0.124803   0.460235   
4  0.710905  0.003597 -0.271783   1.347035  -1.150021 -0.124803  -1.309337   

   REMODEL_Old  REMODEL_Recent  
0     -0.33356       -0.421381  
1     -0.33356        2.372738  
2     -0.33356       -0.421381  
3     -0.33356       -0.421381  
4     -0.33356       -0.421381

In [30]:
# Method 2
# Scaled using the mean and standard deviation
from sklearn.preprocessing import MinMaxScaler, StandardScaler

scaler = StandardScaler()
norm_df = pd.DataFrame(scaler.fit_transform(housing_df), index = housing_df.index,
                      columns = housing_df.columns)
norm_df.head()


C:\Users\DD\Miniconda3\envs\python_finance\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\DD\Miniconda3\envs\python_finance\lib\site-packages\sklearn\base.py:464: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


TOTAL_VALUE       TAX  LOT_SQFT  YR_BUILT  GROSS_AREA  LIVING_AREA  \
0    -0.488921 -0.488550  1.381138 -1.576825   -0.553046    -0.564507   
1     0.200812  0.200806  0.116845  0.229392    0.207214     0.590172   
2    -0.631102 -0.631230  0.457736 -1.298946   -0.713696    -0.529349   
3     1.068020  1.068112  2.807634  0.562847    2.383909     1.759654   
4    -0.616985 -0.616802 -0.478777 -0.743187   -0.627714    -0.405369   

     FLOORS     ROOMS  BEDROOMS  FULL_BATH  HALF_BATH   KITCHEN  FIREPLACE  \
0  0.710966 -0.692039 -0.271806  -0.568577   0.723264 -0.124814  -1.309450   
1  0.710966  2.090505  0.909482   1.347151   0.723264 -0.124814  -1.309450   
2  0.710966  0.699233  0.909482  -0.568577   0.723264 -0.124814  -1.309450   
3 -1.537005  1.394869  2.090769  -0.568577   0.723264 -0.124814   0.460275   
4  0.710966  0.003597 -0.271806   1.347151  -1.150120 -0.124814  -1.309450   

   REMODEL_Old  REMODEL_Recent  
0    -0.333589       -0.421418  
1    -0.333589        2.372943  
2    -0.333589       -0.421418  
3    -0.333589       -0.421418  
4    -0.333589       -0.421418

In [33]:
# Method 3
# Scaled using the Min Max scaler
scaler = MinMaxScaler()

norm_df = pd.DataFrame(scaler.fit_transform(housing_df),
                       index = housing_df.index,
                       columns = housing_df.columns)

norm_df.head()

C:\Users\DD\Miniconda3\envs\python_finance\lib\site-packages\sklearn\preprocessing\data.py:334: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by MinMaxScaler.
  return self.partial_fit(X, y)


TOTAL_VALUE       TAX  LOT_SQFT  YR_BUILT  GROSS_AREA  LIVING_AREA  FLOORS  \
0     0.214953  0.215015  0.197472  0.934858    0.220237     0.177220     0.5   
1     0.276420  0.276448  0.123156  0.967181    0.311878     0.307628     0.5   
2     0.202283  0.202300  0.143194  0.939831    0.200873     0.181191     0.5   
3     0.353702  0.353740  0.281323  0.973148    0.574253     0.439707     0.0   
4     0.203541  0.203586  0.088145  0.949776    0.211237     0.195193     0.5   

      ROOMS  BEDROOMS  FULL_BATH  HALF_BATH  KITCHEN  FIREPLACE  REMODEL_Old  \
0  0.272727     0.250       0.00   0.333333      0.0       0.00          0.0   
1  0.636364     0.375       0.25   0.333333      0.0       0.00          0.0   
2  0.454545     0.375       0.00   0.333333      0.0       0.00          0.0   
3  0.545455     0.500       0.00   0.333333      0.0       0.25          0.0   
4  0.363636     0.250       0.25   0.000000      0.0       0.00          0.0   

   REMODEL_Recent  
0             0.0  
1             1.0  
2             0.0  
3             0.0  
4             0.0

#### Creating the data partition

The standard practice in Machine Learning is to partition the data into 3 parts.

- Training Set
- Validation Set
- Testing Set


The Training set data is used to train our model. We can use the train data to train as many models as we like. The validation partition is used to compare the predictive performance of each model. The test is used once to test the performance of our chosen model. We only test once on the test data to avoid overfitting our model to the data.

In the below code we will partition our data.

In [58]:
from sklearn.model_selection import train_test_split

train_data, temp = train_test_split(housing_df, test_size = 0.5, random_state = 1)
valid_data, test_data = train_test_split(temp, test_size = 0.4, random_state = 1)
print(f"Training Data : {train_data.shape}")
print(f"Validation Data : {valid_data.shape}")
print(f"Test Data : {test_data.shape}")
print()
print(f"Housing Data : {housing_df.shape[0]}")

print(f"Training Data + Validation Data + Test Data = {train_data.shape[0]} + \
{valid_data.shape[0]} + {test_data.shape[0]} = {train_data.shape[0] + valid_data.shape[0] + test_data.shape[0]}")

Training Data : (2901, 15)
Validation Data : (1740, 15)
Test Data : (1161, 15)

Housing Data : 5802
Training Data + Validation Data + Test Data = 2901 + 1740 + 1161 = 5802


#### Building a Predictive Model

- Purpose<br>
Our purpose is to build a model to predict the home values in West Roxbury, Boston
- Obtain data<br>
We will use the 2014 West Roxbury housing data 
- Explore and Clean Data<br>
We will remove the TAX variable, since tax itself is determined by home values. We will also convert the Remodel data to categorical data. We will remove any outliers.
- Reduce data dimension<br>
We will consolidate the Living Area, Room, bedrooms and bathroom(half and full) into smaller number of variables
- Determine the data mining task<br>
We want to predict the Total Value of the house, using the other predictor variables
- Partition the data<br>
We will partition the data into training, validation and prediction sets.
- Choose the technique<br>
In this case we will use the multiple linear regression to predict the home value
- Use the Algorithm<br>
We will use the sklearn.LinearRegression method to predict the house value
- Interpret the result<br>
In this stage we usually try other methods to see how they perform. After choosing the best method(one with the lowest error on the validation data) we use the model to to predict the output on fresh data.
- Deploy the model<br>
After the best model is chosen, we deploy model and score the outcome.

In [69]:
# First load the librarries

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Load the data

housing_df = pd.read_csv('dmba-datasets\dmba\WestRoxbury.csv')

# Rename the columns

housing_df.columns = [x.strip().replace(' ', '_') for x in housing_df.columns]

# Convert the categorical data to dummies

housing_df = pd.get_dummies(housing_df, prefix_sep='_', drop_first=True)

# Exclude the Total value and tax columns

exclude_col = ('TOTAL_VALUE', 'TAX')

# predictors

predictors = [x for x in housing_df.columns if x not in exclude_col]

# outcomes

outcome = 'TOTAL_VALUE'

# Partition data

X = housing_df[predictors]
y = housing_df[outcome]

train_X, valid_X, train_y, valid_y = train_test_split(X, y, 
                                                      test_size = 0.4, random_state=1)

# Training and Fitting the model

model = LinearRegression()
model.fit(train_X, train_y)

# prediction

train_pred = model.predict(train_X)

train_results = pd.DataFrame({'TOTAL_VALUE':train_y,
                            'predicted':train_pred,
                            'residual':train_y - train_pred})

train_results.head(10)


TOTAL_VALUE   predicted   residual
2024        392.0  387.726258   4.273742
5140        476.3  430.785540  45.514460
5259        367.4  384.042952 -16.642952
421         350.3  369.005551 -18.705551
1401        348.1  314.725722  33.374278
2617        528.9  522.662507   6.237493
2083        426.4  386.724566  39.675434
963         351.8  431.046474 -79.246474
1101        344.9  375.749001 -30.849001
3084        270.7  256.686601  14.013399

In [73]:
valid_pred = model.predict(valid_X)
valid_results = pd.DataFrame({'TOTAL_VALUE':valid_y,
                             'predicted':valid_pred,
                             'residual':valid_y - valid_pred})

display(valid_results.head())

TOTAL_VALUE   predicted   residual
1822        462.0  406.946377  55.053623
1998        370.4  362.888928   7.511072
5126        407.4  390.287208  17.112792
808         316.1  382.470203 -66.370203
4034        393.2  434.334998 -41.134998